### PR TITLE
oav support refresh cache per resource provider

### PR DESCRIPTION
**Pull request by ruowan**

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "oav",
-  "version": "0.20.4",
+  "version": "0.20.4-beta.0",
```
Comment (raych1) [inline package.json]: Let's bump minor version for this change.

File: test/utilsTests.ts [hunk 2]
```diff
@@ -43,4 +43,24 @@ describe("Utility functions", () => {
       assert.strictEqual(provider, "Microsoft.Authorization")
     })
   })
+
+  describe("Get provider by swagger file name", () => {
+    it("should return undefined", () => {
+      assert.equal(utils.getProviderBySwaggerFileName(""), undefined)
+    })
+    it("should return resource provider", () => {
+      assert.equal(
```
Comment (raych1) [inline test/utilsTests.ts]: Can we change this test to negative case?

File: lib/validators/liveValidator.ts [hunk 3]
```diff
@@ -825,7 +825,20 @@ export class LiveValidator {
     }
   }
 
-  private async getSwaggerInitializer(swaggerPath: string): Promise<void> {
+  public async refreshResourceProviderCache(resourceProvider: string): Promise<void> {
```
Comment (raych1) [inline lib/validators/liveValidator.ts]: Suggest to pass in string array.

File: lib/validators/liveValidator.ts [hunk 4]
```diff
@@ -825,7 +825,20 @@ export class LiveValidator {
...
+    const specsPaths = (await this.getSwaggerPaths()).filter(
```
Comment (raych1) [inline lib/validators/liveValidator.ts]: Leverage options.swaggerPathsPattern which will be used to match the swagger to load.

File: lib/validators/liveValidator.ts [hunk 5]
```diff
@@ -814,30 +814,42 @@ export class LiveValidator {
...
+        const patterns = pathsPatterns.map(item => path.join(this.options.directory, item))
+        return this.getMatchedPaths(patterns)
       }
     }
   }
 
-  private async getSwaggerInitializer(swaggerPath: string): Promise<void> {
+  public async refreshResourceProviderCache(resourceProviders: string[]): Promise<void> {
+    const pathsPatterns = resourceProviders.map(
+      it => `/specification/**/resource-manager/${it}/**/*.json`
```
Comment (lirenhe) [inline lib/validators/liveValidator.ts]: Could you use a better name to replace 'it'?

File: lib/util/utils.ts [hunk 6]
```diff
@@ -210,6 +210,20 @@ export function setObject(doc: {}, ptr: string, value: unknown, overwrite = true
...
+ * @example input: apimanagement/resource-manager/Microsoft.ApiManagement/preview/2018-01-01/apimwriteonly.json.
+ * return: Microsoft.ApiManagement
+ */
+export function getProviderBySwaggerFileName(filename: string): string | undefined {
+  const pattern = new RegExp("resource-manager/(?<resourceProvider>.*?)/")
+  const matchRes = filename.match(pattern)
+  if (matchRes === null) {
+    return undefined
+  }
+  return matchRes.groups === undefined ? undefined : matchRes.groups.resourceProvider
```
Comment (lirenhe) [inline lib/util/utils.ts]: Could you refine this line?

Comment (ruowan) [reply lib/util/utils.ts]: Use ts3.7 optional chaining.

File: lib/validators/liveValidator.ts (hunk 5 above)
Comment (ruowan) [reply lib/validators/liveValidator.ts]: updated

File: lib/util/utils.ts (hunk 6 above)
Comment (ruowan) [reply lib/util/utils.ts]: We can use optional chaining to instead it after upgrade typescript to 3.7.4 .

File: lib/validators/liveValidator.ts [hunk 7]
```diff
@@ -835,15 +835,33 @@ export class LiveValidator {
...
   public async refreshResourceProviderCache(resourceProviders: string[]): Promise<void> {
     const pathsPatterns = resourceProviders.map(
-      it => `/specification/**/resource-manager/${it}/**/*.json`
+      resourceProvider => `/specification/**/resource-manager/${resourceProvider}/**/*.json`
     )
-    const specsPaths = await this.getSwaggerPaths(pathsPatterns)
-    const promiseFactories = specsPaths.map(swaggerPath => {
-      return this.getSwaggerInitializer(swaggerPath, true)
+    resourceProviders.forEach(resourceProvider => {
+      delete this.cache[resourceProvider.toLowerCase()]
```
Comment (raych1) [inline lib/validators/liveValidator.ts]: The cache item cannot be deleted and has to be updated in place.

Comment (ruowan) [reply lib/validators/liveValidator.ts]: Update equals delete and reload.

**Comment by raych1** [comment]
Close as this is out of date.

**Pull request closed without merging**